{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mapping nucleosomes\n",
    "\n",
    "\n",
    "The goal of the mapping phase is to identify the regions where the nucleosomes are positioned.\n",
    "To do so, we first try to identify the central position of the nucleosome (*dyad*) and\n",
    "then we count 73 bp up and downstream.\n",
    "\n",
    "\n",
    "5 species have been analysed:\n",
    "* [H. sapiens](#sapiens)\n",
    "* [A. thaliana](#thaliana)\n",
    "* [S. cerevisiae](#cerevisiae)\n",
    "* [D. melanogaster](#melanogaster)\n",
    "* [Mus musculus](#musculus)\n",
    "\n",
    "\n",
    "To be able to run this notebook some external data needs to be downloaded. In each section you can find further details."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## H. sapiens <a id=\"sapiens\"></a>\n",
    "\n",
    "\n",
    "Create a **sapiens** folder and place inside:\n",
    "\n",
    "- The 147 bp length mid-fragments MNAse-seq reads which\n",
    "  has been downloaded from http://eqtl.uchicago.edu/nucleosomes/midpoints/mnase_mids_combined_147.wig.gz  \n",
    "Uncompress this file before running the notebook\n",
    "\n",
    "- The file with the hg18 chromosome sizes downloaded from\n",
    "  https://github.com/igvteam/igv/blob/master/genomes/sizes/hg18.chrom.sizes\n",
    "\n",
    "- The hg18 to hg19 chain file downloaded from\n",
    "  http://hgdownload.soe.ucsc.edu/goldenPath/hg18/liftOver/hg18ToHg19.over.chain.gz\n",
    "\n",
    "- The CRG [mappability track](http://rohsdb.cmb.usc.edu/GBshape/cgi-bin/hgFileUi?db=hg19&g=wgEncodeMapability),\n",
    "  that can be downloaded from\n",
    "  http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/wgEncodeCrgMapabilityAlign36mer.bigWig\n",
    "\n",
    "- The positions of the genes downloaded from\n",
    "  ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz  \n",
    "Uncompress this file before running the notebook\n",
    "\n",
    "- The Cancer gene census from COSMIC v81,\n",
    "  You can find information on how to get this version in\n",
    "  https://cancer.sanger.ac.uk/cosmic/help/download#cmd\n",
    "\n",
    "\n",
    "---\n",
    "\n",
    "\n",
    "Two files are generated:\n",
    "- ``dyads.bed.gz`` corresponds to the position of the dyads in mappable non-genic regions\n",
    "- ``dyads_genic.bed.gz`` corresponds to the position of the dyads in genic regions\n",
    "\n",
    "Both correspond to the HG19 genome.\n",
    "\n",
    "---\n",
    "\n",
    "The ``mnase_mids_combined_147.bed.gz`` and ``coverage.bed.gz`` files are used in later analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "source activate env_nucperiod\n",
    "scripts=${PWD}/scripts\n",
    "\n",
    "cd sapiens  # move to workspace\n",
    "\n",
    "wig2bed < mnase_mids_combined_147.wig | gzip > mnase_mids_combined_147.bed.gz\n",
    "# Filter the fragments with 0 reads\n",
    "zcat mnase_mids_combined_147.bed.gz | awk '{OFS=\"\\t\"}{ if($5>0){print $1, $2, $3, $5}}' | \\\n",
    "    gzip > mnase_mids_filtered.bed.gz\n",
    "# Generate a bed file for each chromosome\n",
    "zcat mnase_mids_filtered.bed.gz | awk '{print $0 >> \"mnase_mids_filtered.\"$1\".bed\"}'\n",
    "gzip -f mnase_mids_filtered.*.bed\n",
    "\n",
    "# Compute the triweight kernel smoothing for each chromosome\n",
    "# This part can be easily parallelized by launching one process per file\n",
    "for file in mnase_mids_filtered.*.bed.gz\n",
    "do\n",
    "    f=\"${file/mnase_mids_filtered/mnase_mids_kernel_smoothing}\"\n",
    "    smooth_file=\"${f/.bed./.tsv.}\"\n",
    "    normalize_file=\"${smooth_file/mnase_mids_kernel_smoothing/mnase_mids_kernel_smoothing_normalized}\"\n",
    "    python ${scripts}/kernel_smoothing.py ${file} ${smooth_file}\n",
    "    python ${scripts}/normalize_kernel.py ${smooth_file} ${normalize_file}\n",
    "done\n",
    "\n",
    "\n",
    "#. Concatenate all the chromosome split bed files into a single bed file\n",
    "zcat mnase_mids_kernel_smoothing_normalized.chr1.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr10.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr11.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr12.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr13.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr14.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr15.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr16.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr17.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr18.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr19.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr2.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr20.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr21.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr22.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr3.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr4.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr5.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr6.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr7.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr8.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chr9.tsv.gz \\\n",
    "    mnase_mids_kernel_smoothing_normalized.chrX.tsv.gz \\\n",
    "    > mnase_mids_kernel_smoothing_normalized.bed\n",
    "\n",
    "# Get the positions of the local maxima\n",
    "bedGraphToBigWig mnase_mids_kernel_smoothing_normalized.bed hg18.chrom.sizes \\\n",
    "    mnase_mids_kernel_smoothing_normalized.bw\n",
    "gzip -f mnase_mids_kernel_smoothing_normalized.bed\n",
    "# TODO remove singularity for bwtool\n",
    "singularity run /workspace/projects/bgframework/bgsupport/internal/oriol_bwtool/bwtool.simg find local-extrema -maxima -min-sep=150 mnase_mids_kernel_smoothing_normalized.bw \\\n",
    "    mnase_mids_local_maxima.bed\n",
    "gzip -f mnase_mids_local_maxima.bed\n",
    "# The above command rounds the scores, and we need to get those back\n",
    "intersectBed -a mnase_mids_local_maxima.bed.gz -b mnase_mids_kernel_smoothing_normalized.bed.gz -wo -sorted | \\\n",
    "    awk '{OFS=\"\\t\"}{print $1, $2, $3, $10, $5, $6}' |gzip > mnase_mids_local_maxima.scores.bed.gz\n",
    "\n",
    "# Extend the local maxima 30 and intersect with mid-fragments\n",
    "zcat mnase_mids_local_maxima.scores.bed.gz |\n",
    "    awk '{OFS=\"\\t\";}{print $1, $2-30, $3+30, $1 \"_\" $2 \"_\" $3, $4}' | \\\n",
    "    intersectBed -a mnase_mids_filtered.bed.gz -b stdin -wo | \\\n",
    "    gzip > mnase_mids_local_maxima_intersect.bed.gz\n",
    "\n",
    "# Get the midpoints\n",
    "python ${scripts}/get_dyads.py mnase_mids_local_maxima_intersect.bed.gz \\\n",
    "    mnase_mids_stringency.bed.gz\n",
    "\n",
    "# Lift over the coordinates from hg18 to hg19\n",
    "source deactivate\n",
    "source activate env_crossmap\n",
    "CrossMap.py bed hg18ToHg19.over.chain.gz mnase_mids_stringency.bed.gz mnase_mids_stringency_hg19.bed\n",
    "gzip -f mnase_mids_stringency_hg19.bed\n",
    "source deactivate\n",
    "source activate env_nucperiod\n",
    "\n",
    "# Compute a regions file containing mappable non-genic regions\n",
    "cat gencode.v19.annotation.gtf |grep protein_coding | grep -w gene | \\\n",
    "    awk '{OFS=\"\\t\";} {print $1, $4-501, $5+500}' |gzip > genes_extended.bed.gz\n",
    "# Convert the mappability file to bed\n",
    "bigWigToWig wgEncodeCrgMapabilityAlign36mer.bigWig wgEncodeCrgMapabilityAlign36mer.wig\n",
    "wig2bed --zero-indexed < wgEncodeCrgMapabilityAlign36mer.wig | gzip > wgEncodeCrgMapabilityAlign36mer.bed.gz\n",
    "# Filter unmappable regions\n",
    "zcat wgEncodeCrgMapabilityAlign36mer.bed.gz | awk '$5==1' | gzip > wgEncodeCrgMapabilityAlign36mer_score1.bed.gz\n",
    "# Generate a mappability file without genic regions\n",
    "subtractBed -a wgEncodeCrgMapabilityAlign36mer_score1.bed.gz -b genes_extended.bed.gz |\\\n",
    "    gzip > coverage.bed.gz\n",
    "\n",
    "# Get only nucleosomes in mappable non-genic regions\n",
    "zcat mnase_mids_stringency_hg19.bed.gz | sort -k1,1 -k2,2n | \\\n",
    "    awk '{OFS=\"\\t\";}{print $1, $2-73, $3+73, $4, $5, $6}'  | \\\n",
    "    intersectBed -a stdin -b coverage.bed.gz -f 1 -sorted | \\\n",
    "    awk '{OFS=\"\\t\";}{print $1, $2+73, $3-73, $4, $5, $6}' |  \\\n",
    "    gzip > dyads.bed.gz\n",
    "    \n",
    "# Get the coordinates of the exons\n",
    "grep -w exon gencode.v19.annotation.gtf | grep protein_coding |\\\n",
    "    gzip > gencode.v19.exon_protein_coding.gtf.gz\n",
    "python ${scripts}/exons.py gencode.v19.exon_protein_coding.gtf.gz \\\n",
    "    cancer_gene_census.csv exons.bed.gz\n",
    "mergeBed -i exons.bed.gz | gzip > exons.merged.bed.gz\n",
    "    \n",
    "# Get only nucleosomes falling in genic regions\n",
    "zcat  mnase_mids_stringency_hg19.bed.gz  | \\\n",
    "    awk '{OFS=\"\\t\"}{print $1, $2-58, $3+58}' | \\\n",
    "    intersectBed -a stdin -b exons.merged.bed.gz -f 1 | \\\n",
    "    awk '{OFS=\"\\t\"}{print $1, $2+58, $3-58}' | \\\n",
    "    sort -k1,1 -k2,2n | gzip > dyads_genic.bed.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A. thaliana <a id=\"thaliana\"></a>\n",
    "\n",
    "\n",
    "Create a **thaliana** folder and place inside:\n",
    "\n",
    "- *TAIR10* genome index: can be downloaded from\n",
    "   https://support.illumina.com/sequencing/sequencing_software/igenome.html\n",
    "   (in particular ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Arabidopsis_thaliana/Ensembl/TAIR10/Arabidopsis_thaliana_Ensembl_TAIR10.tar.gz)  \n",
    "You need to uncompress the file\n",
    "\n",
    "- *TAIR 10 chromosome sizes*:\n",
    "  the chromosome sizes for TAIR10 have been downloaded from https://github.com/igvteam/igv/blob/master/genomes/sizes/tair10.chrom.sizes  \n",
    "You have to edit this file to have the chromosomes named as *chrX* instead of the original ``ChrX``.\n",
    "\n",
    "- *TAIR10 gene coordinates*:\n",
    "  download the file from https://www.arabidopsis.org/download_files/Genes/TAIR10_genome_release/TAIR10_gff3/TAIR10_GFF3_genes.gff\n",
    "\n",
    "\n",
    "----\n",
    "\n",
    "The ``dyads.bed.gz`` file contains the positions of the dyads in non genic regions for the TAIR10 genome.\n",
    "\n",
    "---\n",
    "The ``RR1536143_dyads.bed.gz``, ``SRR1536143_dyads_stringency.bed.gz`` and ``coverage.bed.gz`` files are used in later analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "source activate env_nucperiod\n",
    "scripts=${PWD}/scripts\n",
    "\n",
    "cd thaliana  \n",
    "\n",
    "# Download the data RSA: SRR1536143\n",
    "fastq-dump --split-3 -F --read-filter pass SRR1536143\n",
    "\n",
    "# Map to TAIR10 genome\n",
    "bowtie -q --nomaqround --phred33-quals -S --chunkmbs 200 --seed 123 \\\n",
    "    Arabidopsis_thaliana/Ensembl/TAIR10/Sequence/BowtieIndex/genome \\\n",
    "    -1 SRR1536143_pass_1.fastq -2 SRR1536143_pass_2.fastq SRR1536143.sam\n",
    "\n",
    "# Get fragments with size between 146 and 148 bp\n",
    "awk '{OFS=\"\\t\";}{ if (145<$9 && $9<149 && $2==99 && $5>10) {print $3, $4+int($9/2)-1, $4+int($9/2)} }' SRR1536143.sam | \\\n",
    "    sort -k1,1 -k2,2n | uniq -c | \\\n",
    "    awk '{OFS=\"\\t\"} {print \"chr\"$2, $3, $4, $1}' | \\\n",
    "    gzip > SRR1536143_dyads.bed.gz\n",
    "\n",
    "# Repeat the procedure as for H. sapiens:\n",
    "\n",
    "zcat SRR1536143_dyads.bed.gz | awk '{print $0 >> \"SRR1536143_dyads.\"$1\".bed\"}'\n",
    "gzip -f SRR1536143_dyads.*.bed\n",
    "for file in SRR1536143_dyads.*.bed.gz\n",
    "do\n",
    "    f=\"${file/SRR1536143_dyads/SRR1536143_dyads_kernel_smoothing}\"\n",
    "    smooth_file=\"${f/.bed./.tsv.}\"\n",
    "    normalize_file=\"${smooth_file/SRR1536143_dyads_kernel_smoothing/SRR1536143_dyads_kernel_smoothing_normalized}\"\n",
    "    python ${scripts}/kernel_smoothing.py ${file} ${smooth_file}\n",
    "    python ${scripts}/normalize_kernel.py ${smooth_file} ${normalize_file}\n",
    "done\n",
    "zcat SRR1536143_dyads_kernel_smoothing_normalized.chr1.tsv.gz \\\n",
    "    SRR1536143_dyads_kernel_smoothing_normalized.chr2.tsv.gz \\\n",
    "    SRR1536143_dyads_kernel_smoothing_normalized.chr3.tsv.gz \\\n",
    "    SRR1536143_dyads_kernel_smoothing_normalized.chr4.tsv.gz \\\n",
    "    SRR1536143_dyads_kernel_smoothing_normalized.chr5.tsv.gz \\\n",
    "    > SRR1536143_dyads_kernel_smoothing_normalized.bed\n",
    "bedGraphToBigWig SRR1536143_dyads_kernel_smoothing_normalized.bed tair10.chrom.sizes \\\n",
    "    SRR1536143_dyads_kernel_smoothing_normalized.bw\n",
    "gzip -f SRR1536143_dyads_kernel_smoothing_normalized.bed\n",
    "\n",
    "# TODO remove singularity for bwtool\n",
    "singularity run /workspace/projects/bgframework/bgsupport/internal/oriol_bwtool/bwtool.simg find local-extrema -maxima -min-sep=150 SRR1536143_dyads_kernel_smoothing_normalized.bw \\\n",
    "    SRR1536143_dyads_local_maxima.bed\n",
    "gzip -f SRR1536143_dyads_local_maxima.bed\n",
    "intersectBed -a SRR1536143_dyads_local_maxima.bed.gz -b SRR1536143_dyads_kernel_smoothing_normalized.bed.gz \\\n",
    "    -wo -sorted | awk '{OFS=\"\\t\"}{print $1, $2, $3, $10, $5, $6}' | \\\n",
    "    gzip > SRR1536143_dyads_local_maxima.scores.bed.gz\n",
    "zcat SRR1536143_dyads_local_maxima.scores.bed.gz | \\\n",
    "    awk '{OFS=\"\\t\";}{print $1, $2-30, $3+30, $1 \"_\" $2 \"_\" $3, $4}' | \\\n",
    "    intersectBed -a SRR1536143_dyads.bed.gz -b stdin -wo | \\\n",
    "    gzip > SRR1536143_dyads_local_maxima_intersect.bed.gz\n",
    "python ${scripts}/get_dyads.py SRR1536143_dyads_local_maxima_intersect.bed.gz \\\n",
    "    SRR1536143_dyads_stringency.bed.gz\n",
    "\n",
    "# Get the TAIR10 CDS coordinates\n",
    "awk '{OFS=\"\\t\";}{if ($3==\"CDS\") {print tolower($1), $4, $5}}' TAIR10_GFF3_genes.gff |gzip > TAIR10_CDS.bed.gz\n",
    "\n",
    "# Extend the maxima 73 bp on each direcction and remove maximas in genes\n",
    "zcat SRR1536143_dyads_stringency.bed.gz | \\\n",
    "    awk '{OFS=\"\\t\";}{print $1, $2-73, $3+73, $4, $5, $6}'  | \\\n",
    "    intersectBed -a stdin -b TAIR10_CDS.bed.gz -v | \\\n",
    "    awk '{OFS=\"\\t\";}{print $1, $2+73, $3-73, $4, $5, $6}' | \\\n",
    "    sort -k1,1 -k2,2n | gzip > dyads.bed.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## S. cerevisiae <a id=\"cerevisiae\"></a>\n",
    "\n",
    "Create a **cerevisiae** folder and place inside:\n",
    "\n",
    "- Data downloaded from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3786739/bin/NIHMS370046-supplement-5.txt\n",
    "\n",
    "- *sacCer2ToSacCer3.over.chain.gz*: can be downloaded from\n",
    "  http://hgdownload.soe.ucsc.edu/goldenPath/sacCer2/liftOver/\n",
    "\n",
    "- saccer3 chromosome sizes downloaded from: hgdownload.cse.ucsc.edu/goldenPath/sacCer3/bigZips/sacCer3.chrom.sizes\n",
    "\n",
    "\n",
    "----\n",
    "\n",
    "The ``dyads.bed.gz`` contains the positions of the dyads in the sacCer3 genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "source activate env_nucperiod\n",
    "scripts=${PWD}/scripts\n",
    "\n",
    "cd cerevisiae\n",
    "\n",
    "# Create a bed file\n",
    "awk '{OFS=\"\\t\";}{print $1, $2-1, $2, $3, $4 }' NIHMS370046-supplement-5.txt |\\\n",
    "    sort -k1,1 -k2,2n | gzip > cerevisiae.bed.gz\n",
    "\n",
    "# Compute distances\n",
    "closestBed -a cerevisiae.bed.gz -b cerevisiae.bed.gz -io -d | gzip > cerevisiae_closest.bed.gz\n",
    "\n",
    "# Remove overlapping dyads\n",
    "python ${scripts}/remove_closest.py cerevisiae_closest.bed.gz cerevisiae_nooverlap.bed.gz\n",
    "\n",
    "# Perform a lift over from saceer2 to saccer 3\n",
    "source deactivate\n",
    "source activate env_crossmap\n",
    "CrossMap.py bed sacCer2ToSacCer3.over.chain.gz cerevisiae_nooverlap.bed.gz dyads_all.bed\n",
    "gzip -f dyads_all.bed\n",
    "source deactivate\n",
    "source activate env_nucperiod\n",
    "\n",
    "python ${scripts}/filter_extremes.py dyads_all.bed.gz sacCer3.chrom.sizes dyads.bed.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## D. melanogaster <a id='melanogaster'></a>\n",
    "\n",
    "\n",
    "Create a **melanogaster** folder and place inside the data downloaded from\n",
    "https://doi.org/10.1371/journal.pgen.1004457.s018 as ``intergenic.txt`` and\n",
    "https://doi.org/10.1371/journal.pgen.1004457.s019 as ``intronic.txt``\n",
    "\n",
    "----\n",
    "\n",
    "The ``dyads.bed.gz`` file contains the positions of the dyads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from intervaltree import IntervalTree\n",
    "\n",
    "np.random.seed(123)\n",
    "\n",
    "df1 = pd.read_csv('melanogaster/intergenic.txt', sep ='\\t', names = ['chr', 's-1', 's'], skiprows=1)\n",
    "df2 = pd.read_csv('melanogaster/intronic.txt', sep ='\\t', names = ['chr', 's-1', 's'], skiprows=1)\n",
    "df = pd.concat([df1, df2])\n",
    "rows = np.random.choice(df.index.values, len(df), replace=False)\n",
    "sampled_df = df.iloc[rows]\n",
    "\n",
    "tree = IntervalTree()\n",
    "\n",
    "# Perform a random sampling\n",
    "output = 'melanogaster/dyads.bed.gz'\n",
    "with gzip.open(output, 'wt') as outfile:\n",
    "    for chrom, data in df.groupby(by='chr'):\n",
    "        set_forbidden = set()\n",
    "        for i, row in data.iterrows():\n",
    "            center = row['s-1'] + 73 + 50\n",
    "            if not tree.search(center-73, center+73):\n",
    "                out = '{}\\t{}\\t{}\\n'.format(row['chr'], center-1, center)\n",
    "                outfile.write(out)\n",
    "                tree.addi(center - 73, center + 73)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mus musculus <a id='musculus'></a>\n",
    "\n",
    "Create a **musculus** folder and place inside:\n",
    "\n",
    "- The data downloaded from\n",
    "  https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM2183909&format=file&file=GSM2183909%5Funique%2Emap%5F95pc%2Etxt%2Egz  \n",
    "Uncompress this file before running the notebook\n",
    "\n",
    "- Mouse genome from\n",
    "  ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M1/gencode.vM1.annotation.gtf.gz  \n",
    "Uncompress this file before running the notebook\n",
    "\n",
    "- The mappability track\n",
    "  that can be downloaded from\n",
    "  http://hgdownload.cse.ucsc.edu/goldenPath/mm9/encodeDCC/wgEncodeMapability/wgEncodeCrgMapabilityAlign36mer.bigWig\n",
    "\n",
    "\n",
    "----\n",
    "\n",
    "\n",
    "The ``dyads.bed.gz`` file corresponds to the position of the dyads in mappable non-genic regions of the mm9 genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "source activate env_nucperiod\n",
    "scripts=${PWD}/scripts\n",
    "\n",
    "cd musculus\n",
    "\n",
    "# Get the extended positions of the genes\n",
    "cat gencode.vM1.annotation.gtf |grep protein_coding | grep -w gene | \\\n",
    "    awk '{OFS=\"\\t\";} {print $1, $4-501, $5+500}' |gzip > genes_extended.bed.gz\n",
    "# Convert the mappability file to bed\n",
    "bigWigToWig wgEncodeCrgMapabilityAlign36mer.bigWig wgEncodeCrgMapabilityAlign36mer.wig\n",
    "wig2bed --zero-indexed < wgEncodeCrgMapabilityAlign36mer.wig | gzip > wgEncodeCrgMapabilityAlign36mer.bed.gz\n",
    "# Filter unmappable regions\n",
    "zcat wgEncodeCrgMapabilityAlign36mer.bed.gz | awk '$5==1' |\\\n",
    "    gzip > wgEncodeCrgMapabilityAlign36mer_score1.bed.gz\n",
    "    \n",
    "# Generate a mappability file without genic regions\n",
    "subtractBed -a wgEncodeCrgMapabilityAlign36mer_score1.bed.gz -b genes_extended.bed.gz |\\\n",
    "    gzip > coverage.bed.gz\n",
    "\n",
    "# Remove genic and unmappable regions\n",
    "cat GSM2183909_unique.map_95pc.txt | awk '{OFS=\"\\t\";}{if ($1==20){print \"chrX\", $2-1, $2}else if ($1==21){print \"chrY\", $2-1, $2} else if ($1==22){} else {print \"chr\"$1, $2-1, $2}}' | \\\n",
    "    sort -k1,1 -k2,2n | \\\n",
    "    awk '{OFS=\"\\t\";}{print $1, $2-73, $3+73}'  | \\\n",
    "    intersectBed -a stdin -b coverage.bed.gz -f 1 -sorted | \\\n",
    "    awk '{OFS=\"\\t\";}{print $1, $2+73, $3-73}' |  \\\n",
    "    gzip > dyads.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:env_nucperiod]",
   "language": "python",
   "name": "conda-env-env_nucperiod-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
