{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome-wide periodicity analysis\n",
    "\n",
    "\n",
    "The goal of the periodicty phase is assess the periodicity of the $WW$ mofitf ($WW = \\{AA, AT, TA, TT\\}$) in the categorical signal defined by an input DNA sequence.\n",
    "\n",
    "Note, this analysis generated a huge amount of intermediate data. Thus, we are deleting most of them and kept only the esential files. If you would like to check all the intermediate files, please change the temporary folder for another."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data\n",
    "\n",
    "To be able to run this analysis you need to download some external datasets.\n",
    "Datasets consist on genome DNA sequences (fasta files) downloaded from [UCSC genome browser](https://genome.ucsc.edu/) and [ENSEMBL genome browse](http://ensemblgenomes.org/)\n",
    "\n",
    "For analysing the data, you should create a separate folder for each genome you want to analyse. The name of the folder must be: ``<family>_<species>``. If the species names has more than one word, separate them with ``_``; e.g.: ``insects_apis_mellifera``.\n",
    "\n",
    "For ensembl genomes the files named as ``.dna_rm.toplevel.fa.gz`` have been used and for UCSC we have downloaded the ``.fa.masked.gz`` files.\n",
    "\n",
    "If there is only one file inside, it is assumed that it contains all the chromosomes/groups from the whole genome analysis. If there are many files, it is assume that there is already one file per chromosome/group.\n",
    "\n",
    "Please, extract *tar* files, but compress all fasta files in *gzip* format."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Periodicity analysis for all data (we only keep intermediate files generated for S. cerevisiae)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "source activate env_nucperiod\n",
    "cores=6\n",
    "\n",
    "for organism_folder in *_*\n",
    "do\n",
    "\n",
    "    if [ -f \"${organism_folder}\" ]\n",
    "    then\n",
    "        continue\n",
    "    elif [ \"${organism_folder}\" == \"fungi_saccharomyces_cerevisiae\" ]\n",
    "    then\n",
    "        tmp_folder=${organism_folder}\n",
    "    else\n",
    "        tmp_folder=`mktemp -d`\n",
    "    fi\n",
    "    \n",
    "    chr_folder=${tmp_folder}/chroms\n",
    "    mkdir -p ${chr_folder}\n",
    "    chunks_folder=${tmp_folder}/fasta_chunks\n",
    "    mkdir -p ${chunks_folder}\n",
    "    rand_chunks_folder=${tmp_folder}/random_fasta_chunks\n",
    "    mkdir -p ${rand_chunks_folder}\n",
    "    motif_folder=${tmp_folder}/motif_counts\n",
    "    mkdir -p ${motif_folder}\n",
    "    rand_motif_folder=${tmp_folder}/random_motif_counts\n",
    "    mkdir -p ${rand_motif_folder}\n",
    "\n",
    "    # Generate the input\n",
    "    nfiles=$(ls ${organism_folder}/*.fa.*gz | wc -l)\n",
    "    if [ $nfiles = 1 ]\n",
    "    then\n",
    "        # Split the file into one file per CHR\n",
    "        zcat ${organism_folder}/*.fa.*gz | \\\n",
    "            awk -v outdir=\"${chr_folder}/\" '/^>/{s=substr($0,2); split(s, res, \" \"); t=res[1] \".fa\"} {print > (outdir t)}'         \n",
    "    else\n",
    "        # Assume it is already splited and just uncompress the files\n",
    "        for file in ${organism_folder}/*.fa.*gz\n",
    "        do\n",
    "            name=$(basename ${file})\n",
    "            name=${name/.fa/}\n",
    "            name=${name/.gz/}\n",
    "            name=${name/.masked/}\n",
    "            zcat ${file} > ${chr_folder}/${name}.fa\n",
    "        done\n",
    "    fi\n",
    "\n",
    "    # Delete all files with size less than 1M\n",
    "    find ${chr_folder} -name \"*.fa\" -not -size +1M -delete\n",
    "\n",
    "    # Create the chunck for each fasta file per chromosome\n",
    "    for file in ${chr_folder}/*.fa\n",
    "    do\n",
    "        chr=$(basename $file)\n",
    "        chr=${chr/.fa/}\n",
    "        mkdir -p ${chunks_folder}/${chr}\n",
    "        faSplit size ${file} 500000 ${chunks_folder}/${chr}/ \\\n",
    "            -lift=\"${chunks_folder}/${chr}/${chr}.lft\" -extra=500000 -verbose=0\n",
    "    done\n",
    "\n",
    "    # Remove small files\n",
    "    for file in ${chunks_folder}/*/*.fa\n",
    "    do\n",
    "        if [ -f ${file} ]\n",
    "        then\n",
    "            letters=`cat ${file} | wc -c`\n",
    "            if [ ${letters} -lt 1000000 ]\n",
    "            then\n",
    "                rm ${file}\n",
    "            fi\n",
    "        fi\n",
    "    done\n",
    "\n",
    "    # Generate the Markov-models from the chunks\n",
    "    for file in ${chunks_folder}/*/*.fa\n",
    "    do\n",
    "        if [ -f ${file} ]\n",
    "        then\n",
    "            chr=$(basename $(dirname ${file}))\n",
    "            name=$(basename ${file})\n",
    "            name=${name%.fa}\n",
    "            output_dir=${rand_chunks_folder}/${chr}/${name}\n",
    "            mkdir -p ${output_dir}\n",
    "            python scripts/shuffle.py \\\n",
    "                --chunk ${file} --output_folder ${output_dir} \\\n",
    "                --repeats 100 --cores ${cores}\n",
    "        fi\n",
    "    done\n",
    "\n",
    "    # Compute motif counts\n",
    "    for dir in $(find ${chunks_folder} -maxdepth 1 -mindepth 1 -type d)\n",
    "    do\n",
    "        chr=$(basename ${dir})\n",
    "        output=${motif_folder}/${chr}\n",
    "        mkdir -p ${output}\n",
    "        python scripts/correlation.py ${dir} ${output} --cores ${cores} --wildcard \"*.fa\"\n",
    "    done\n",
    "\n",
    "    # Compute motif counts for random chunks\n",
    "    for dir in $(find ${rand_chunks_folder} -maxdepth 1 -mindepth 1 -type d)\n",
    "    do\n",
    "        chr=$(basename ${dir})\n",
    "        for subdir in $(find ${dir} -maxdepth 1 -mindepth 1 -type d)\n",
    "        do\n",
    "            name=$(basename ${subdir})\n",
    "            output=${rand_motif_folder}/${chr}/${name}\n",
    "            mkdir -p ${output}\n",
    "            python scripts/correlation.py ${subdir} ${output} --cores ${cores} --wildcard \"*.txt.gz\"\n",
    "        done\n",
    "    done\n",
    "\n",
    "    # Compute the power and SNR for all the chunks\n",
    "    python scripts/spectral.py ${motif_folder} ${organism_folder}/observed_spectra.npz --cores ${cores} \\\n",
    "        --table ${organism_folder}/observed_chunks.tsv\n",
    "\n",
    "    # Same for randomizations\n",
    "    python scripts/spectral.py ${rand_motif_folder} ${organism_folder}/random_spectra.npz --cores ${cores} \\\n",
    "            --table ${organism_folder}/random_chunks.tsv --randomized\n",
    "\n",
    "    # Build summary and metrics\n",
    "    python scripts/summary.py  ${organism_folder}/observed_chunks.tsv ${organism_folder}/random_chunks.tsv \\\n",
    "        ${organism_folder}/summary.tsv ${organism_folder}/metrics.tsv\n",
    "\n",
    "    # Clean up\n",
    "    if [ \"${tmp_folder}\" != \"${organism_folder}\" ]\n",
    "    then\n",
    "        rm -r ${tmp_folder}\n",
    "    fi\n",
    "    \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:env_nucperiod]",
   "language": "python",
   "name": "conda-env-env_nucperiod-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
